File: android/core/src/test/kotlin/com/theformofvoid/voidbridge/core/ProtocolTest.kt
package com.theformofvoid.voidbridge.core

import java.util.concurrent.CopyOnWriteArrayList
import kotlin.test.AfterTest
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlin.test.fail

class ProtocolTest {
    // Same vectors as TestKnownAnswers in internal/protocol (Go).
    @Test
    fun knownAnswersMatchGo() {
        val code = Keys(Protocol.masterFromCode("ABCD-EFGH-IJKL-MNOP"))
        assertEquals("9e3f6a5f684ceef721f4e5f99e8e61f435dc7f28aaa0c4658803808be73c7d59", code.master.toHex())
        val acct = Keys(Protocol.masterFromAccount(" Alice ", "correct horse"))
        assertEquals("02de1a5fe7b5eaaac9fea0798fd73520a99acb964edfecc937f4433ff5a6715a", acct.master.toHex())
        assertEquals("7115bab6f55f9e7149e3eedae05a2f4cbb3977d0b1e2ca52306f821e73b96c2b", code.link.toHex())
        assertEquals("3baec9577d1694f354d68114a6fa44fe4cfabfdddede1548844c58018e6ce2e3", code.content.toHex())
        assertEquals("5d880d9c8233fb24a0819091fdc4fb1cde7878be0f4a8cebb774d55f8961615e", acct.auth.toHex())
        assertEquals("96001ca5c8506c14", code.fingerprint)
        val sk = Protocol.sessionKey(code.link, ByteArray(16) { 1 }, ByteArray(16) { 2 })
        assertEquals("7b364ea83e6e69c4ceb6c712569e0614f6628a08ae1c4d8f8bd88871599ea039", sk.toHex())
        assertEquals("57a9fc60d3c90b7001a2379a1334f0ccd51dbce163248160116646cfcf", FrameCipher(sk, false).seal("""{"t":"ready"}""".toByteArray()).toHex())
        assertEquals("0000000c7b2274223a2270696e67227d09", Message(Message.PING, body = byteArrayOf(9)).encode().toHex())
    }

    @Test
    fun codes() {
        assertEquals("ABCDEFOI234567B", Protocol.normalizeCode("abcd-ef01-2345-6789"))
        assertTrue(Protocol.validCode(Protocol.newGroupCode()))
        assertEquals("ABCD-EFGH-IJKL-MNOP", Protocol.formatCode("abcdefghijklmnop"))
    }

    @Test
    fun clipSealing() {
        val k = Keys(Protocol.masterFromCode("AAAA-BBBB-CCCC-DDDD"))
        val m = Protocol.sealClip(k.content, Message(Message.CLIP, id = "x", origin = "o", time = 5, clipType = "text", mime = "text/plain"), "secret".toByteArray())
        assertContentEquals("secret".toByteArray(), Protocol.openClip(k.content, m))
        assertNull(Protocol.openClip(k.content, m.copy(time = 6)), "tampered header accepted")
        val d = Message.decode(m.encode())
        assertContentEquals(m.body, d.body)
        assertEquals("o", d.origin)
    }

    @Test
    fun cipherRejectsReplay() {
        val sk = Protocol.sessionKey(ByteArray(32), Protocol.newNonce(), Protocol.newNonce())
        val cli = FrameCipher(sk, false)
        val srv = FrameCipher(sk, true)
        val f = cli.seal("one".toByteArray())
        assertContentEquals("one".toByteArray(), srv.open(f))
        assertFailsWith<AuthException> { srv.open(f) }
    }

    // ---- Kotlin devices talking to each other ----

    class Dev(name: String, keys: Keys) : Node.Listener {
        val got = CopyOnWriteArrayList<Content>()
        val node = Node(Identity("kt-$name", name), keys, this)
        val peers = PeerManager(node, keys, port = 0, discovery = false).apply {
            allowLoopback = true
            dialIntervalMs = 50
            pingIntervalMs = 200
            idleTimeoutMs = 2_000
            start()
        }
        val addr get() = "127.0.0.1:${peers.listenPort}"
        override fun apply(content: Content, from: Message) { got.add(content) }
        fun lastText() = got.lastOrNull()?.text
        fun stop() { peers.stop(); node.shutdown() }
    }

    private val cleanup = mutableListOf<() -> Unit>()

    @AfterTest
    fun tearDown() = cleanup.forEach { it() }

    private fun dev(name: String, keys: Keys) = Dev(name, keys).also { d -> cleanup.add { d.stop() } }

    private fun eventually(ms: Long = 10_000, what: String = "condition", cond: () -> Boolean) {
        val end = System.currentTimeMillis() + ms
        while (System.currentTimeMillis() < end) {
            if (cond()) return
            Thread.sleep(20)
        }
        fail("$what never became true")
    }

    private val group = Keys(Protocol.masterFromCode("AAAA-BBBB-CCCC-DDDD"))

    @Test
    fun threeDevicesChainAndImages() {
        val a = dev("a", group)
        val b = dev("b", group)
        val c = dev("c", group)
        a.peers.setManual(listOf(b.addr))
        b.peers.setManual(listOf(c.addr))
        eventually(what = "links") { a.node.devices().isNotEmpty() && c.node.devices().isNotEmpty() }
        a.node.localCopy(Content(Protocol.CLIP_TEXT, text = "from a"))
        eventually(what = "b and c got it") { b.lastText() == "from a" && c.lastText() == "from a" }
        val img = ByteArray(3 shl 20) { (it % 251).toByte() }
        c.node.localCopy(Content(Protocol.CLIP_IMAGE, data = img, mime = "image/png"))
        eventually(what = "image reached a") { a.got.lastOrNull()?.data?.contentEquals(img) == true }
    }

    @Test
    fun sensitiveNotSentAndOtherGroupIgnored() {
        val a = dev("a", group)
        val b = dev("b", group)
        val x = dev("x", Keys(Protocol.masterFromCode("ZZZZ-ZZZZ-ZZZZ-ZZZZ")))
        a.peers.setManual(listOf(b.addr, x.addr))
        eventually(what = "a-b link") { a.node.devices().size == 1 }
        a.node.localCopy(Content(Protocol.CLIP_TEXT, text = "hunter2", sensitive = true))
        a.node.localCopy(Content(Protocol.CLIP_TEXT, text = "fine"))
        eventually(what = "b got fine") { b.lastText() == "fine" }
        assertTrue(b.got.none { it.text == "hunter2" })
        assertTrue(x.got.isEmpty())
        assertEquals(0, x.node.devices().size)
    }

    // ---- interop with the Go implementation (tools/interop) ----

    private fun interopEnv(): Pair<String, String>? {
        val peer = System.getenv("VOIDBRIDGE_INTEROP_PEER") ?: return null
        val code = System.getenv("VOIDBRIDGE_INTEROP_CODE") ?: return null
        return peer to code
    }

    @Test
    fun interopDirectLinkWithGo() {
        val (peerAddr, code) = interopEnv() ?: return
        val d = dev("kt-direct", Keys(Protocol.masterFromCode(code)))
        d.peers.setManual(listOf(peerAddr))
        eventually(what = "linked to Go peer") { d.node.devices().any { it.id == "go-peer" } }
        d.node.localCopy(Content(Protocol.CLIP_TEXT, text = "ping:direct ✓"))
        eventually(what = "pong from Go") { d.lastText() == "pong:direct ✓" }
        val img = ByteArray(1 shl 20) { (it * 7).toByte() }
        d.node.localCopy(Content(Protocol.CLIP_IMAGE, data = img, mime = "image/png"))
        val want = "got-image:" + java.security.MessageDigest.getInstance("SHA-256").digest(img).toHex()
        eventually(what = "Go got the image") { d.lastText() == want }
    }

    @Test
    fun interopServerWithGo() {
        val server = System.getenv("VOIDBRIDGE_INTEROP_SERVER") ?: return
        val base = ServerApi.normalizeUrl(server)
        val keys = Keys(Protocol.masterFromAccount("alice", "interop-password"))
        val me = Identity("kt-account", "Kotlin phone")
        val api = ServerApi(base)
        assertTrue(api.info().hasUsers)
        assertFailsWith<ServerException> { ServerApi(base).login("alice", Keys(Protocol.masterFromAccount("alice", "wrong")), me) }
        val login = api.login("alice", keys, me)
        assertNotNull(login.token)

        val got = CopyOnWriteArrayList<String>()
        val node = Node(me, keys, object : Node.Listener {
            override fun apply(content: Content, from: Message) { got.add(content.text) }
        })
        val relay = RelayClient(base, login.token, node, object : RelayClient.Listener {
            override fun onState(connected: Boolean, error: String?) { if (connected) connectedFlag = true }
            override fun onUnauthorized() {}
        })
        relay.start()
        cleanup.add { relay.stop(); node.shutdown() }
        eventually(what = "server connection and Go device visible") { connectedFlag && node.devices().any { it.id == "go-account" } }
        node.localCopy(Content(Protocol.CLIP_TEXT, text = "ping:server"))
        eventually(what = "pong through server") { got.lastOrNull() == "pong:server" }
    }

    @Volatile private var connectedFlag = false

    @Test
    fun normalizeUrl() {
        assertEquals("http://pi:47831", ServerApi.normalizeUrl("pi"))
        assertEquals("http://pi:9000", ServerApi.normalizeUrl("pi:9000/"))
        assertEquals("https://vb.example.com", ServerApi.normalizeUrl("https://vb.example.com"))
    }
}

File: internal/protocol/protocol_test.go
package protocol

import (
	"bytes"
	"encoding/hex"
	"errors"
	"net"
	"testing"
	"time"
)

func TestCodes(t *testing.T) {
	c := NewGroupCode()
	if len(c) != 19 || !ValidCode(c) {
		t.Fatalf("bad code %q", c)
	}
	if NormalizeCode("abcd-ef01-2345-6789") != "ABCDEFOI234567B" {
		t.Fatal(NormalizeCode("abcd-ef01-2345-6789"))
	}
	if FormatCode("abcdefghijklmnop") != "ABCD-EFGH-IJKL-MNOP" {
		t.Fatal(FormatCode("abcdefghijklmnop"))
	}
}

// Shared with android/core ProtocolTest.knownAnswersMatchGo.
func TestKnownAnswers(t *testing.T) {
	check := func(name string, got []byte, want string) {
		t.Helper()
		if hex.EncodeToString(got) != want {
			t.Errorf("%s = %x, want %s", name, got, want)
		}
	}
	code := KeysFromMaster(MasterFromCode("ABCD-EFGH-IJKL-MNOP"))
	check("code master", code.Master, "9e3f6a5f684ceef721f4e5f99e8e61f435dc7f28aaa0c4658803808be73c7d59")
	acct := KeysFromMaster(MasterFromAccount(" Alice ", "correct horse"))
	check("account master", acct.Master, KA["account master"])
	check("link", code.Link, KA["link"])
	check("content", code.Content, KA["content"])
	check("auth", acct.Auth, KA["auth"])
	if fp := code.Fingerprint(); fp != KA["fingerprint"] {
		t.Errorf("fingerprint = %s", fp)
	}
	sk := SessionKey(code.Link, bytes.Repeat([]byte{1}, 16), bytes.Repeat([]byte{2}, 16))
	check("session", sk, KA["session"])
	check("first client frame", NewCipher(sk, false).Seal([]byte(`{"t":"ready"}`)), KA["frame"])
	m := &Message{Type: TypeClip, ID: "c1", Origin: "d1", Time: 1700000000000, ClipType: ClipText, Mime: "text/plain"}
	if got := string(clipAAD(m)); got != "voidbridge-clip-v2|c1|d1|1700000000000|text|text/plain" {
		t.Errorf("aad = %s", got)
	}
	check("encoded frame", Encode(&Message{Type: TypePing, Body: []byte{9}}), "0000000c7b2274223a2270696e67227d09")
}

// KA holds the known answers (hex) for values not easy to write inline.
var KA = map[string]string{
	"account master": "02de1a5fe7b5eaaac9fea0798fd73520a99acb964edfecc937f4433ff5a6715a",
	"link":           "7115bab6f55f9e7149e3eedae05a2f4cbb3977d0b1e2ca52306f821e73b96c2b",
	"content":        "3baec9577d1694f354d68114a6fa44fe4cfabfdddede1548844c58018e6ce2e3",
	"auth":           "5d880d9c8233fb24a0819091fdc4fb1cde7878be0f4a8cebb774d55f8961615e",
	"fingerprint":    "96001ca5c8506c14",
	"session":        "7b364ea83e6e69c4ceb6c712569e0614f6628a08ae1c4d8f8bd88871599ea039",
	"frame":          "57a9fc60d3c90b7001a2379a1334f0ccd51dbce163248160116646cfcf",
}

func TestClipSealing(t *testing.T) {
	k := KeysFromMaster(MasterFromCode("AAAA-BBBB-CCCC-DDDD"))
	m := &Message{Type: TypeClip, ID: "x", Origin: "o", Time: 5, ClipType: ClipText, Mime: "text/plain"}
	SealClip(k.Content, m, []byte("secret"))
	if p, err := OpenClip(k.Content, m); err != nil || string(p) != "secret" {
		t.Fatal(p, err)
	}
	m.Time = 6 // tampered header
	if _, err := OpenClip(k.Content, m); !errors.Is(err, ErrDecrypt) {
		t.Fatal("tampered header accepted")
	}
	m.Time = 5
	other := KeysFromMaster(MasterFromCode("AAAA-BBBB-CCCC-DDDE"))
	if _, err := OpenClip(other.Content, m); !errors.Is(err, ErrDecrypt) {
		t.Fatal("wrong key accepted")
	}
	// Round trip through a frame.
	d, err := Decode(Encode(m))
	if err != nil || !bytes.Equal(d.Body, m.Body) || d.Origin != "o" {
		t.Fatal(d, err)
	}
}

func TestCipherReplay(t *testing.T) {
	sk := SessionKey(make([]byte, 32), NewNonce(), NewNonce())
	cli, srv := NewCipher(sk, false), NewCipher(sk, true)
	f := cli.Seal([]byte("one"))
	if p, err := srv.Open(f); err != nil || string(p) != "one" {
		t.Fatal(p, err)
	}
	if _, err := srv.Open(f); err == nil {
		t.Fatal("replay accepted")
	}
	if _, err := srv.Open(srv.Seal([]byte("x"))); err == nil {
		t.Fatal("reflection accepted")
	}
}

func pipe(t *testing.T) (net.Conn, net.Conn) {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	defer ln.Close()
	ch := make(chan net.Conn)
	go func() { c, _ := ln.Accept(); ch <- c }()
	a, err := net.Dial("tcp", ln.Addr().String())
	if err != nil {
		t.Fatal(err)
	}
	return a, <-ch
}

func TestHandshakeAndBigFrame(t *testing.T) {
	k := KeysFromMaster(MasterFromCode("AAAA-BBBB-CCCC-DDDD"))
	a, b := pipe(t)
	ch := make(chan *Conn)
	go func() {
		c, err := Handshake(b, k.Link, Identity{ID: "pc", Name: "PC", Kind: "windows"}, []string{"10.0.0.2:47829"}, false, 5*time.Second)
		if err != nil {
			t.Error(err)
		}
		ch <- c
	}()
	cli, err := Handshake(a, k.Link, Identity{ID: "phone", Name: "Phone", Kind: "android"}, nil, true, 5*time.Second)
	if err != nil {
		t.Fatal(err)
	}
	srv := <-ch
	if cli.Peer.ID != "pc" || srv.Peer.Kind != "android" || cli.PeerAddrs[0] != "10.0.0.2:47829" {
		t.Fatal("identities not exchanged")
	}
	big := bytes.Repeat([]byte("x"), 10<<20)
	go cli.Send(&Message{Type: TypeClip, ID: "i", Body: big})
	m, err := srv.Recv(5 * time.Second)
	if err != nil || !bytes.Equal(m.Body, big) {
		t.Fatal(err)
	}
}

func TestHandshakeWrongGroup(t *testing.T) {
	a, b := pipe(t)
	ch := make(chan error)
	go func() {
		_, err := Handshake(b, KeysFromMaster(MasterFromCode("AAAA-BBBB-CCCC-DDDD")).Link, Identity{ID: "pc"}, nil, false, 5*time.Second)
		ch <- err
	}()
	_, err := Handshake(a, KeysFromMaster(MasterFromCode("AAAA-BBBB-CCCC-DDDE")).Link, Identity{ID: "phone"}, nil, true, 5*time.Second)
	if !errors.Is(err, ErrAuth) {
		t.Fatalf("dialer: %v", err)
	}
	if err := <-ch; !errors.Is(err, ErrAuth) {
		t.Fatalf("listener: %v", err)
	}
}
